sl

```diff
@@ -1,27 +1,105 @@
+    exit
+fi
 
-TRUFFLE_LIB=./graalvm/jre/lib/truffle
+#######################################################################
+#                 Check graalvm location and version.                 #
+#      Fallback is JAVA_HOME which may or may not be a GraalVM.       #
+#######################################################################
+if [[ -d "graalvm" ]]; then
+    GRAALVM_VERSION=$(grep "GRAALVM_VERSION" ./graalvm/release)
+    if [[ "$GRAALVM_VERSION" == "" ]]; then
+       echo "Cannot find GRAALVM_VERSION entry in ./graalvm/release"
+       exit
+    fi
+    GRAALVM_VERSION=$(echo "$GRAALVM_VERSION" | awk 'BEGIN {FS="\""} {print $2}')
+    if [[ "$GRAALVM_VERSION" != "$VERSION" ]]; then
+        echo "Wrong version of GraalVM in ./graalvm. Expected: $VERSION, found $GRAALVM_VERSION"
+        exit
+    fi
+    JAVACMD=${JAVACMD:=./graalvm/bin/java}
+elif [[ "$JAVA_HOME" != "" ]]; then
+    GRAALVM_VERSION=$(grep "GRAALVM_VERSION" "$JAVA_HOME"/release)
+    if [[ "$GRAALVM_VERSION" != "" ]]; then
+        GRAALVM_VERSION=$(echo "$GRAALVM_VERSION" | awk 'BEGIN {FS="\""} {print $2}')
+        if [[ "$GRAALVM_VERSION" != "$VERSION" ]]; then
+            echo "Wrong version of GraalVM in \$JAVA_HOME. Expected: $VERSION, found $GRAALVM_VERSION"
+            exit
+        fi
+    fi
+    JAVACMD=${JAVACMD:=$JAVA_HOME/bin/java}
+else
+    echo "The ./graalvm directory missing and JAVA_HOME not set"
+    exit
+fi
 
-$JAVACMD $JAVA_ARGS -XX:-UseJVMCIClassLoader -Xbootclasspath/a:$TRUFFLE_LIB/truffle-api.jar:$TRUFFLE_LIB/locator.jar:$TRUFFLE_LIB/truffle-nfi.jar -cp ./target/classes com.oracle.truffle.sl.SLMain $PROGRAM_ARGS
+#######################################################################
+#          Parse arguments, prepare Java command and execute.         #
+#######################################################################
+if [[ "$GRAALVM_VERSION" != "" ]]; then
+    PROGRAM_ARGS=""
+    JAVA_ARGS=""
+
+    for opt in "$@"
+    do
+      case $opt in
+        -debug)
+          JAVA_ARGS="$JAVA_ARGS -Xdebug -Xrunjdwp:transport=dt_socket,server=y,address=8000,suspend=y" ;;
+        -dump)
+          JAVA_ARGS="$JAVA_ARGS -Dgraal.Dump= -Dgraal.MethodFilter=Truffle.* -Dgraal.TruffleBackgroundCompilation=false -Dgraal.TraceTruffleCompilation=true -Dgraal.TraceTruffleCompilationDetails=true" ;;
+        -disassemble)
+          JAVA_ARGS="$JAVA_ARGS -XX:CompileCommand=print,*OptimizedCallTarget.callRoot -XX:CompileCommand=exclude,*OptimizedCallTarget.callRoot -Dgraal.TruffleBackgroundCompilation=false -Dgraal.TraceTruffleCompilation=true -Dgraal.TraceTruffleCompilationDetails=true" ;;
+        -J*)
+          opt=${opt:2}
+          JAVA_ARGS="$JAVA_ARGS $opt" ;;
+        *)
+          PROGRAM_ARGS="$PROGRAM_ARGS $opt" ;;
+      esac
+    done
+    $JAVACMD $JAVA_ARGS -Dtruffle.class.path.append=$LANGUAGE_PATH -cp $LAUNCHER_PATH $MAIN_CLASS $PROGRAM_ARGS
+else
+    echo "NOTE: Could not find GraalVM, using $JAVA_HOME"
```

Comment:
Improve error:
`Warning: Could not find GraalVM on $JAVA_HOME or in ./graalvm folder. Running on JDK without support for compilation.`